### PR TITLE
[Encoding] introduce encoding component

### PR DESCRIPTION
**Pull request by azjezz**
- [x] base 64
- [x] base 16 ( hex )

--- 

Changelog:

Features
- added `Psl\Encoding\Base64\encode`
- added `Psl\Encoding\Base64\decode`
- added `Psl\Encoding\Hex\encode`
- added `Psl\Encoding\Hex\decode`

### CODE REVIEW

File: src/Psl/Encoding/Base64/decode.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+function decode(string $base64): string
+{
+    if (!preg_match('%^[a-zA-Z0-9/+]*={0,2}$%', $base64)) {
+        throw new Exception\RangeException(
+            'The given base64 string contains characters outside the base64 range.'
+        );
+    }
+
+    return (string) @base64_decode($base64, true);
```
Comment (azjezz) [inline src/Psl/Encoding/Base64/decode.php]: 🤔

Comment (veewee) [reply src/Psl/Encoding/Base64/decode.php]: When does it trigger a warning (or something similar)?
I've tried with some utf8 chars, but only got `false` due to invalid range.

Comment (azjezz) [reply src/Psl/Encoding/Base64/decode.php]: incorrect padding, but not sure, I would need to add more tests I guess.

Comment (azjezz) [reply src/Psl/Encoding/Base64/decode.php]: hm, actually, it returns false without any warning for invalid padding. so I guess we just check the return type, if it's false, we throw `Exception\PaddingException`.

Comment (azjezz) [reply src/Psl/Encoding/Base64/decode.php]: or we can check the length before calling base64_decode, as base64 encoded strings length is a multiple of 4 when padded correctly.

Comment (veewee) [reply src/Psl/Encoding/Base64/decode.php]: Probably best to do both? Check the length + the return type

**Pull request merged**